Snowplow/SPDefaultNetworkConnection.m

```diff
@@ -0,0 +1,205 @@
+        _httpMethod = SPRequestPost;
+        _protocol = SPHttps;
+        _emitThreadPoolSize = 15;
+        _byteLimitGet = 40000;
+        _byteLimitPost = 40000;
+        _customPostPath = nil;
+        _dataOperationQueue = [[NSOperationQueue alloc] init];
+        _builderFinished = NO;
+    }
+    return self;
+}
+
+- (void) setup {
+    _dataOperationQueue.maxConcurrentOperationCount = _emitThreadPoolSize;
+    NSString *urlPrefix = _protocol == SPHttp ? @"http://" : @"https://";
+    NSString *urlSuffix = _httpMethod == SPRequestGet ? kSPEndpointGet : kSPEndpointPost;
+    if (_customPostPath && _httpMethod == SPRequestPost) {
+        urlSuffix = _customPostPath;
+    }
+    _urlEndpoint = [NSURL URLWithString:[NSString stringWithFormat:@"%@%@%@", urlPrefix, _urlString, urlSuffix]];
+    
+    if ([_urlEndpoint scheme] && [_urlEndpoint host]) {
+        SPLogDebug(@"Emitter URL created successfully '%@'", _urlEndpoint);
+        NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
+        [userDefaults setObject:_urlString forKey:kSPErrorTrackerUrl];
+        [userDefaults setObject:urlSuffix forKey:kSPErrorTrackerProtocol];
+        [userDefaults setObject:urlPrefix forKey:kSPErrorTrackerMethod];
+    } else {
+        SPLogDebug(@"Invalid emitter URL: '%@'", _urlEndpoint);
+        NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
+        [userDefaults setObject:@"acme.com" forKey:kSPErrorTrackerUrl];
+        [userDefaults setObject:kSPEndpointPost forKey:kSPErrorTrackerProtocol];
+        [userDefaults setObject:@"http://" forKey:kSPErrorTrackerMethod];
+    }
+    _builderFinished = YES;
+}
+
+// Required
+
+- (void)setUrlEndpoint:(NSString *)urlEndpoint {
+    _urlString = urlEndpoint;
+    if (_builderFinished) {
+        [self setup];
+    }
+}
+
+- (void)setHttpMethod:(SPRequestOptions)method {
+    _httpMethod = method;
+    if (_builderFinished && _urlEndpoint != nil) {
+        [self setup];
+    }
+}
+
+- (void)setProtocol:(SPProtocol)protocol {
+    _protocol = protocol;
+    if (_builderFinished && _urlEndpoint != nil) {
+        [self setup];
+    }
+}
+
+- (void)setEmitThreadPoolSize:(NSUInteger)emitThreadPoolSize {
+    _emitThreadPoolSize = emitThreadPoolSize;
+    if (_dataOperationQueue.maxConcurrentOperationCount != emitThreadPoolSize) {
+        _dataOperationQueue.maxConcurrentOperationCount = _emitThreadPoolSize;
+    }
+}
+
+- (void)setByteLimitGet:(NSUInteger)byteLimitGet {
+    _byteLimitGet = byteLimitGet;
+}
+
+- (void)setByteLimitPost:(NSUInteger)byteLimitPost {
+    _byteLimitPost = byteLimitPost;
+}
+
+- (void)setCustomPostPath:(NSString *)customPath {
+    _customPostPath = customPath;
+}
+
+// MARK: - Implement SPNetworkConnection protocol
+
+- (SPRequestOptions)httpMethod {
+    return _httpMethod;
+}
+
+- (NSURL *)url {
+    return _urlEndpoint.copy;
+}
+
+- (NSArray<SPRequestResult *> *)sendRequests:(NSArray<SPRequest *> *)requests {
```

Comment:
I think it is out of scope for this release but it would be good to find a way remove the blocking nature of this event sending and move to a better asynchronous pattern. Something to think about for v2.

Reply:
Yep definitely!